File: src/lib/embeddings.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export type EmbeddingProviderKind = "local" | "remote";

// Mirrors `domain::embeddings::ResolvedEmbeddingConfig` — what
// `embedding_get_config` returns (bundled preset merged with overrides).
export type ResolvedEmbeddingConfig = {
  kind: EmbeddingProviderKind;
  remoteBaseUrl: string | null;
  remoteModel: string | null;
  remoteDimensions: number | null;
  remoteTrustedCertPem: string | null;
  remoteSystemId: string | null;
  remoteDisableTlsVerification: boolean;
};

// Mirrors `domain::embeddings::EmbeddingProviderConfig` — the settings-layer
// override persisted by `embedding_set_config`. `kind: null` means inherit
// from the bundled preset.
export type EmbeddingProviderConfig = {
  kind: EmbeddingProviderKind | null;
  remoteBaseUrl: string | null;
  remoteModel: string | null;
  remoteDimensions: number | null;
  remoteTrustedCertPem: string | null;
  remoteSystemId: string | null;
  remoteDisableTlsVerification: boolean | null;
};

// Mirrors `domain::embeddings::ModelStatus` (adjacently tagged,
// `#[serde(tag = "status")]`).
export type ModelStatus =
  | { status: "notDownloaded" }
  | { status: "downloading"; progress: number }
  | { status: "ready" }
  | { status: "error"; message: string };

export type SyncStats = {
  embedded: number;
  skippedUnchanged: number;
  removed: number;
};

// Mirrors `domain::embeddings::EmbeddingIndexStatus`.
export type EmbeddingIndexStatus = {
  synced: boolean;
  embeddedCount: number;
  /** Persisted index exists but predates a version bump (or a different
   * index root) — left untouched on disk, not loaded; needs a real sync to
   * repair, distinct from "never synced". */
  stale: boolean;
  /** Files still queued for background processing (see `SyncTrigger`'s
   * `"background"` value) — `0` when nothing's deferred, which is most
   * syncs (a small non-doc change set is folded into the synchronous pass
   * instead). Non-zero after a fresh project's first sync, or after a
   * routine sync catches a large upstream change; documentation itself is
   * always prioritized ahead of this backlog, on every sync. The index
   * itself is safe to use in the meantime, just incomplete outside
   * `docs_root`. */
  backgroundPending: number;
};

// Mirrors `commands::repo_index::RepoIndexSummary` — a live snapshot of
// `RepositoryIndex`/`ChunkIndex`'s current resident state (no walk, no I/O,
// safe to call any time; reports whatever the last sync left resident, or
// all-zero before the first one).
export type RepoIndexSummary = {
  filesIndexed: number;
  byLanguage: Record<string, number>;
  chunksIndexed: number;
};

export type ModelDownloadProgress = {
  progress: number;
  error?: string;
  cancelled?: boolean;
};

// Mirrors the Rust `SyncPhase`/`SyncProgressPayload` emitted by
// `commands::embeddings::embedding_sync`.
export type SyncPhase = "chunking" | "embedding";

// Mirrors the Rust `SyncTrigger` — distinguishes a full, user-triggered
// `embedding_sync` from a file-watcher-driven incremental per-file tick,
// and from the low-priority backlog catch-up that follows a fresh
// project's first sync.
export type SyncTrigger = "full" | "incremental" | "background";

export type SyncProgress = {
  phase: SyncPhase;
  current: number;
  total: number;
  trigger: SyncTrigger;
};

export function getEmbeddingConfig(): Promise<ResolvedEmbeddingConfig> {
  return invoke<ResolvedEmbeddingConfig>("embedding_get_config");
}

export function setEmbeddingConfig(config: EmbeddingProviderConfig): Promise<void> {
  return invoke("embedding_set_config", { config });
}

/** Write-only — there is no `getEmbeddingRemoteApiKey`, only a status check. */
export function setEmbeddingRemoteApiKey(apiKey: string): Promise<void> {
  return invoke("embedding_set_remote_api_key", { apiKey });
}

export function hasEmbeddingRemoteApiKey(): Promise<boolean> {
  return invoke<boolean>("embedding_has_remote_api_key");
}

export function getEmbeddingModelStatus(): Promise<ModelStatus> {
  return invoke<ModelStatus>("embedding_model_status");
}

/** Resolves once the (potentially multi-minute) download finishes or fails
 * — subscribe via `listenModelDownloadProgress` for live status meanwhile. */
export function downloadEmbeddingModel(): Promise<void> {
  return invoke("embedding_download_model");
}

/** The underlying blocking download has no interrupt hook, so this can't
 * stop in-flight network I/O — it only tells the backend (and the UI) to
 * stop trusting whatever that attempt eventually reports back. */
export function cancelEmbeddingModelDownload(): Promise<void> {
  return invoke("embedding_cancel_model_download");
}

/** Rebuilds the repo/chunk index for the current project and reconciles
 * embeddings against it (new chunk → embed, changed hash → re-embed,
 * deleted chunk → drop its vector). */
export function syncEmbeddings(): Promise<SyncStats> {
  return invoke<SyncStats>("embedding_sync");
}

/** Read-only: is the current project's index already built? Backed by the
 * persisted/resident `EmbeddingIndex` itself, not by whether `syncEmbeddings`
 * happened to run earlier in this session — safe to call on every mount to
 * recover real state after a remount. */
export function getEmbeddingIndexStatus(): Promise<EmbeddingIndexStatus> {
  return invoke<EmbeddingIndexStatus>("embedding_index_status");
}

/** Read-only, cheap (in-memory only): the current project's `RepositoryIndex`/
 * `ChunkIndex` state as of the last sync — per-language file counts and a
 * chunk count, previously computed on every full sync and silently
 * discarded (`services::repo_index::RepoIndexStats`), now exposed directly. */
export function getRepoIndexSummary(): Promise<RepoIndexSummary> {
  return invoke<RepoIndexSummary>("repo_index_summary");
}

export function listenModelDownloadProgress(
  onProgress: (payload: ModelDownloadProgress) => void,
): Promise<UnlistenFn> {
  return listen<ModelDownloadProgress>("embedding:model-download-progress", (event) =>
    onProgress(event.payload),
  );
}

/** Fires repeatedly while a `syncEmbeddings()` call is in flight — first
 * `phase: "chunking"` (re-chunking changed files), then `phase: "embedding"`
 * (calling the provider in batches). Not emitted at all for a no-op sync
 * (nothing changed since the last one). */
export function listenSyncProgress(
  onProgress: (payload: SyncProgress) => void,
): Promise<UnlistenFn> {
  return listen<SyncProgress>("embedding:sync-progress", (event) => onProgress(event.payload));
}

/** Stops the file-watcher-driven incremental sync, if one is running for
 * the current project. Call when a project closes without a new one
 * opening in the same session — otherwise the backend's own attach logic
 * naturally swaps the watcher to whichever project opens next. */
export function teardownIncrementalWatcher(): Promise<void> {
  return invoke("embedding_index_teardown");
}

/** Tells the backend which files are open in the editor, for `syncEmbeddings`
 * to prioritize on a fresh project's first sync (open files + their direct
 * includes/xrefs get chunked+embedded before the rest of the repo). Pass
 * `EditorTab.path` values verbatim — already relative to the project's docs
 * root, exactly what the backend expects; no conversion needed here. */
export function setEmbeddingPriorityFiles(relativePaths: string[]): Promise<void> {
  return invoke("embedding_set_priority_files", { relativePaths });
}

File: src-tauri/src/domain/embeddings.rs
//! Types for the embedding layer built on top of the Chunk Index:
//! `Chunk -> Embedding -> vector index`. This module knows nothing about
//! `fastembed`, ONNX, `usearch`, or HTTP — those are `infra` concerns
//! implementing `EmbeddingProvider`/the vector store against these types.

use thiserror::Error;

/// A dense embedding vector. BGE-M3 (the local provider) produces 1024
/// dimensions; a remote provider may differ — `EmbeddingProvider::dimensions`
/// is how a caller finds out which, rather than assuming 1024 everywhere.
#[derive(Debug, Clone, PartialEq)]
pub struct Embedding(pub Vec<f32>);

/// What `EmbeddingIndex` stores per chunk. `chunk_hash` — not the chunk's
/// text — is the staleness signal: `EmbeddingIndex::sync` re-embeds a chunk
/// only when this no longer matches the chunk's current
/// `ChunkMetadata::hash`. Deliberately does **not** carry the vector itself
/// — that already lives in `VectorStore`/`usearch::Index`, the only place
/// anything actually reads it back (via `EmbeddingIndex::search`); keeping
/// a second copy here would just double the resident memory cost of every
/// embedded chunk for no functional benefit.
#[derive(Debug, Clone, PartialEq)]
pub struct EmbeddingRecord {
    pub chunk_hash: blake3::Hash,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum EmbeddingProviderKind {
    Local,
    Remote,
}

impl Default for EmbeddingProviderKind {
    fn default() -> Self {
        Self::Local
    }
}

/// One entry from the compiled-in system providers manifest's top-level
/// `embedding` section (see `infra::embedding_provider_manifest`) — a
/// global default, independent of any LLM provider id. Deserialize-only;
/// every field is `Option` so a fork can leave the template as explicit
/// `null`s (meaning "use the Local BGE-M3 provider") and fill them later
/// without touching Rust.
#[derive(Debug, Clone, Default, PartialEq, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EmbeddingPreset {
    #[serde(default)]
    pub base_url: Option<String>,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub dimensions: Option<usize>,
    #[serde(default)]
    pub trusted_cert_pem: Option<String>,
    /// Alfa internal API: required `systemId` header on `/embeddings`.
    #[serde(default)]
    pub system_id: Option<String>,
    /// Skip TLS certificate validation — sandbox-only escape hatch when the
    /// server leaf cert is expired but the endpoint is still reachable.
    #[serde(default)]
    pub disable_tls_verification: Option<bool>,
}

/// Persisted globally (`AppSettings.embedding`) as an **override** layer on
/// top of the bundled `EmbeddingPreset` — same "field wins when `Some`,
/// `None` means inherit" shape as `LlmProviderConfig`. The remote API key
/// is deliberately **not** a field here: it goes through
/// `infra::embedding_credentials_store` (encrypted), never through plain
/// `settings.json`.
#[derive(Debug, Clone, Default, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EmbeddingProviderConfig {
    /// `None` = inherit from the bundled preset (Remote when the preset
    /// has both `baseUrl` and `model`, else Local). Explicit `Some(Local)` /
    /// `Some(Remote)` pins the choice regardless of the preset.
    #[serde(default)]
    pub kind: Option<EmbeddingProviderKind>,
    #[serde(default)]
    pub remote_base_url: Option<String>,
    #[serde(default)]
    pub remote_model: Option<String>,
    /// `usearch`'s index needs a fixed dimension count at construction
    /// time, and unlike the local BGE-M3 provider (always 1024), a remote
    /// service's dimension count depends entirely on which model it's
    /// running — there's no way to discover it without either an extra API
    /// round-trip or the user stating it up front. Settings asks for this
    /// when Remote is selected; `None` falls back to the preset, then to
    /// `DEFAULT_REMOTE_DIMENSIONS` (OpenAI's `text-embedding-3-small`
    /// size, the most common default).
    #[serde(default)]
    pub remote_dimensions: Option<usize>,
    #[serde(default)]
    pub remote_trusted_cert_pem: Option<String>,
    #[serde(default)]
    pub remote_system_id: Option<String>,
    #[serde(default)]
    pub remote_disable_tls_verification: Option<bool>,
}

/// Merged, ready-to-use view of the embedding provider — bundled preset
/// folded with the settings-layer override. What
/// `services::embedding_config::resolve_embedding_config` produces and what
/// runtime / IPC actually consume. Serialize-only.
#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedEmbeddingConfig {
    pub kind: EmbeddingProviderKind,
    pub remote_base_url: Option<String>,
    pub remote_model: Option<String>,
    pub remote_dimensions: Option<usize>,
    pub remote_trusted_cert_pem: Option<String>,
    pub remote_system_id: Option<String>,
    pub remote_disable_tls_verification: bool,
}

impl Default for ResolvedEmbeddingConfig {
    fn default() -> Self {
        Self {
            kind: EmbeddingProviderKind::Local,
            remote_base_url: None,
            remote_model: None,
            remote_dimensions: None,
            remote_trusted_cert_pem: None,
            remote_system_id: None,
            remote_disable_tls_verification: false,
        }
    }
}

/// Fallback when neither the override nor the preset sets dimensions.
pub const DEFAULT_REMOTE_DIMENSIONS: usize = 1536;

/// Local model download/readiness state. `Downloading` only carries a
/// meaningful `progress` if the download path used one of `fastembed`'s
/// coarse-grained hooks — see `services::embedding_model` for the current
/// caveat on how fine-grained this actually is.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(tag = "status", rename_all = "camelCase")]
pub enum ModelStatus {
    NotDownloaded,
    Downloading { progress: f32 },
    Ready,
    Error { message: String },
}

#[derive(Debug, Error)]
pub enum EmbeddingError {
    #[error("embedding provider error: {0}")]
    Provider(String),
    #[error("http error: {0}")]
    Http(String),
    #[error("tls configuration error: {0}")]
    Tls(String),
    #[error("vector store error: {0}")]
    VectorStore(String),
    #[error("io error: {0}")]
    Io(#[source] std::io::Error),
    #[error("{0}")]
    Message(String),
}

/// One embedding backend — local on-device inference or a remote HTTP API,
/// selected by `ResolvedEmbeddingConfig.kind`. Synchronous (not `async fn`):
/// both concrete implementations are naturally blocking (`fastembed`'s
/// inference has no Tokio dependency; the remote provider uses a blocking
/// HTTP client rather than expanding this project's minimal `tokio`
/// features), so callers run this inside `spawn_blocking` — the same
/// pattern `services::standards::check_standards` and
/// `services::ai_tools::execute_tool` already use on the IPC boundary.
pub trait EmbeddingProvider: Send + Sync {
    /// Batched — callers embed every pending chunk in one call, not one
    /// call per chunk.
    fn embed(&self, texts: &[&str]) -> Result<Vec<Embedding>, EmbeddingError>;
    fn dimensions(&self) -> usize;
}

#[derive(Debug, Clone, Default, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SyncStats {
    pub embedded: usize,
    pub skipped_unchanged: usize,
    pub removed: usize,
}

/// Read-only "is this project's index built" signal — derived from the
/// persisted/resident `EmbeddingIndex` itself (`embedded_count`), not from
/// whether a `sync()` happened to run earlier in this process's lifetime.
/// Lets the UI show real state on mount instead of resetting to "not yet
/// synced" every time a component remounts.
#[derive(Debug, Clone, Default, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EmbeddingIndexStatus {
    pub synced: bool,
    pub embedded_count: usize,
    /// The persisted index exists but predates a `CHUNK_VERSION`/
    /// `INDEX_VERSION` bump (or was built for a different `index_root`) —
    /// `services::index_store_ensure` detected the mismatch but, being a
    /// read-only attach, left the on-disk data untouched rather than
    /// wiping it. `synced` is `false` in this case too (nothing was
    /// loaded), but the reason is different from "never synced": a real
    /// `embedding_sync` will repair and rebuild it, not start from a blank
    /// slate.
    pub stale: bool,
    /// Files still queued for `run_background_backlog_sync` — `0` when
    /// nothing's deferred (the common case: `embedding_sync` folds a small
    /// non-doc change set inline rather than deferring it, see
    /// `commands::embeddings::split_sync_tiers`). Non-zero after a fresh
    /// project's first sync, or after a routine sync catches a large
    /// upstream change (e.g. a big `git pull`) — either way this can be
    /// nonzero on more than just a project's very first sync now that
    /// documentation is prioritized on every sync, not just the first.
    /// Read from `commands::embeddings::BackgroundBacklogSlot`, not a
    /// hand-maintained counter, so it survives an app restart or a panicked
    /// background task without drifting.
    pub background_pending: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_defaults_to_inherit_with_no_remote_fields() {
        let config = EmbeddingProviderConfig::default();
        assert_eq!(config.kind, None);
        assert_eq!(config.remote_base_url, None);
        assert_eq!(config.remote_dimensions, None);
        assert_eq!(config.remote_trusted_cert_pem, None);
    }

    #[test]
    fn deserializes_legacy_config_kind_as_explicit_override() {
        let config: EmbeddingProviderConfig =
            serde_json::from_str(r#"{"kind":"remote","remoteBaseUrl":"https://x"}"#).unwrap();
        assert_eq!(config.kind, Some(EmbeddingProviderKind::Remote));
        assert_eq!(config.remote_base_url.as_deref(), Some("https://x"));
        assert_eq!(config.remote_dimensions, None);
    }

    #[test]
    fn deserializes_legacy_config_without_kind_as_inherit() {
        let config: EmbeddingProviderConfig =
            serde_json::from_str(r#"{"remoteBaseUrl":"https://x"}"#).unwrap();
        assert_eq!(config.kind, None);
        assert_eq!(config.remote_base_url.as_deref(), Some("https://x"));
    }

    #[test]
    fn preset_deserializes_explicit_nulls() {
        let preset: EmbeddingPreset = serde_json::from_str(
            r#"{"baseUrl":null,"model":null,"dimensions":null,"trustedCertPem":null}"#,
        )
        .unwrap();
        assert_eq!(preset.base_url, None);
        assert_eq!(preset.model, None);
        assert_eq!(preset.dimensions, None);
        assert_eq!(preset.trusted_cert_pem, None);
    }

    #[test]
    fn model_status_json_shape_is_tagged_by_status() {
        let json = serde_json::to_string(&ModelStatus::Downloading { progress: 0.5 }).unwrap();
        assert_eq!(json, r#"{"status":"downloading","progress":0.5}"#);
    }
}
